#### 1. Importing some important libraries

In [3]:
import numpy as np
import pandas as pd
import ast

#### 2. Reading datasets

In [4]:
movies = pd.read_csv('tmdb_5000_movies.csv')
credits = pd.read_csv('tmdb_5000_credits.csv')

In [5]:
movies.head(5)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

In [6]:
credits.head(5)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

#### 3. Merging both datasets and removing similar columns for better understanding

In [7]:
movies = movies.merge(credits, on = 'title')

In [8]:
movies.shape  #one column is reduced

(4809, 23)

In [9]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4809 non-null   int64  
 1   genres                4809 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4809 non-null   int64  
 4   keywords              4809 non-null   object 
 5   original_language     4809 non-null   object 
 6   original_title        4809 non-null   object 
 7   overview              4806 non-null   object 
 8   popularity            4809 non-null   float64
 9   production_companies  4809 non-null   object 
 10  production_countries  4809 non-null   object 
 11  release_date          4808 non-null   object 
 12  revenue               4809 non-null   int64  
 13  runtime               4807 non-null   float64
 14  spoken_languages      4809 non-null   object 
 15  status               

#### 4. Removing unneccesary columns which are not helful for algorithm

Columns :
    * budget   <br>
    * homepge   <br>
    * original_language  <br>
    * popularity   <br>
    * production_companies  <br>
    * production_countries  <br>
    * revenue  <br>
    * runtime  <br>
    * spoken_languages  <br>
    * status (Released or Not)  <br>
    * tagline   <br>
    * vote_average   <br>
    * vote_count   <br>
    * movie_id  <br>
    

In [10]:
movies  = movies.drop(['budget','homepage','original_language','popularity','production_companies','production_countries','revenue','runtime','spoken_languages','status','tagline','vote_average','vote_count','movie_id'], axis = 1)

In [11]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4809 entries, 0 to 4808
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   genres          4809 non-null   object
 1   id              4809 non-null   int64 
 2   keywords        4809 non-null   object
 3   original_title  4809 non-null   object
 4   overview        4806 non-null   object
 5   release_date    4808 non-null   object
 6   title           4809 non-null   object
 7   cast            4809 non-null   object
 8   crew            4809 non-null   object
dtypes: int64(1), object(8)
memory usage: 375.7+ KB


In [12]:
# rearranging columns
movies = movies[['id','original_title','overview','genres','keywords','cast','crew','release_date']]

In [13]:
movies.head()

id                            original_title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew release_date  
0  [{"credit_id": "52fe48009251416c750aca23", "de...   2009-12-10  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...   2007-05-19  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...   2015-10-26  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...   2012-07-16  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...   2012-03-07

#### 5. Handling missing values

In [14]:
movies.isnull().sum()

id                0
original_title    0
overview          3
genres            0
keywords          0
cast              0
crew              0
release_date      1
dtype: int64

In [15]:
movies.dropna(inplace=True)

In [16]:
movies.isnull().sum()

id                0
original_title    0
overview          0
genres            0
keywords          0
cast              0
crew              0
release_date      0
dtype: int64

#### 6. Handling duplicate values

In [17]:
movies.duplicated().sum()

0

#### 7. Data Preprocessing

In [18]:
movies.iloc[0].genres

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

* converting 1 --> 2

1. : '[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'
2.  : ['Action','Adventure','Fantasy','SciFi']

In [19]:
def convert(dic):
    l = list()
    for i in ast.literal_eval(dic):
        l.append(i['name'])
    return l

In [20]:
movies['genres'] = movies['genres'].apply(convert)
movies['genres']

0       [Action, Adventure, Fantasy, Science Fiction]
1                        [Adventure, Fantasy, Action]
2                          [Action, Adventure, Crime]
3                    [Action, Crime, Drama, Thriller]
4                [Action, Adventure, Science Fiction]
                            ...                      
4804                        [Action, Crime, Thriller]
4805                                [Comedy, Romance]
4806               [Comedy, Drama, Romance, TV Movie]
4807                                               []
4808                                    [Documentary]
Name: genres, Length: 4805, dtype: object

In [21]:
movies.head()

id                            original_title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew release_date  
0  [{"credit_id": "52fe48009251416c750aca23", "de...   2009-12-10  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...   2007-05-19  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...   2015-10-26  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...   2012-07-16  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...   2012-03-07

In [22]:
movies['keywords'] = movies['keywords'].apply(convert)

In [23]:
movies.head()

id                            original_title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew release_date  
0  [{"credit_id": "52fe48009251416c750aca23", "de...   2009-12-10  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...   2007-05-19  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...   2015-10-26  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...   2012-07-16  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...   2012-03-07

* cast : We need only top 3 characters

In [24]:
def convert_cast(dic):
    l = list()
    counter = 0
    for i in ast.literal_eval(dic):
        if counter != 3 :
            l.append(i['name'])
            counter += 1
        else:
            break
    return l

In [25]:
movies['cast'] = movies['cast'].apply(convert_cast)

In [26]:
movies.head()

id                            original_title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]   
3      [Christian Bale, Michael Caine, Gary Oldman]   
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]   

                                                crew release_date  
0  [{"credit_id": "52fe48009251416c750aca23", "de...   2009-12-10  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...   2007-05-19  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...   2015-10-26  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...   2012-07-16  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...   2012-03-07

In [27]:
movies['crew'][0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

* We need only Director name

In [28]:
def fetch_director(dic):
    l = list()
    for i in ast.literal_eval(dic):
        if i['job'] == 'Director':
            l.append(i['name'])
            break
    return l

In [29]:
movies['crew'] = movies['crew'].apply(fetch_director)
movies['crew']

0           [James Cameron]
1          [Gore Verbinski]
2              [Sam Mendes]
3       [Christopher Nolan]
4          [Andrew Stanton]
               ...         
4804     [Robert Rodriguez]
4805         [Edward Burns]
4806          [Scott Smith]
4807          [Daniel Hsia]
4808     [Brian Herzlinger]
Name: crew, Length: 4805, dtype: object

In [30]:
movies.head()

id                            original_title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                               cast                 crew  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]   
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]   
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]   

  release_date  
0   2009-12-10  
1   2007-05-19  
2   2015-10-26  
3   2012-07-16  
4   2012-03-07

In [31]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4805 entries, 0 to 4808
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   id              4805 non-null   int64 
 1   original_title  4805 non-null   object
 2   overview        4805 non-null   object
 3   genres          4805 non-null   object
 4   keywords        4805 non-null   object
 5   cast            4805 non-null   object
 6   crew            4805 non-null   object
 7   release_date    4805 non-null   object
dtypes: int64(1), object(7)
memory usage: 466.9+ KB


* before merging columns we need to convert overview columns from <br>     string --> list form

In [32]:
movies['overview'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

In [33]:
movies['overview'] = movies['overview'].apply(lambda x:x.split())
movies['overview'][0]

['In',
 'the',
 '22nd',
 'century,',
 'a',
 'paraplegic',
 'Marine',
 'is',
 'dispatched',
 'to',
 'the',
 'moon',
 'Pandora',
 'on',
 'a',
 'unique',
 'mission,',
 'but',
 'becomes',
 'torn',
 'between',
 'following',
 'orders',
 'and',
 'protecting',
 'an',
 'alien',
 'civilization.']

In [34]:
movies.head()

id                            original_title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                               cast                 crew  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]   
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]   
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]   

  release_date  
0   2009-12-10  
1   2007-05-19  
2   2015-10-26  
3   2012-07-16  
4   2012-03-07

Now we have to remove space b/w names because model will read it like two tags : <br>

* ex : 'Shivam Pal ---> 'ShivamPal'

In [35]:
movies['genres'] = movies['genres'].apply(lambda x: [i.replace(" ","") for i in x])
movies['keywords'] = movies['keywords'].apply(lambda x: [i.replace(" ","") for i in x])
movies['cast'] = movies['cast'].apply(lambda x: [i.replace(" ","") for i in x])
movies['crew'] = movies['crew'].apply(lambda x: [i.replace(" ","") for i in x])

In [36]:
movies.head()

id                            original_title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            cast                crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]   
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]   
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]   

  release_date  
0   2009-12-10  
1   2007-05-19  
2   2015-10-26  
3   2012-07-16  
4   2012-03-07

 * Now we need to concanate 'overview','genres','keywords','cast','crew' columns in a single column as a tags

In [37]:
movies['tags'] = movies['overview'] + movies['genres'] + movies['overview'] + movies['cast'] + movies['overview'] + movies['crew']

In [38]:
movies.head()

id                            original_title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            cast                crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]   
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]   
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]   

  release_date                                               tags  
0   2009-12-10  [In, the, 22nd, century,, a, paraplegic, Marin...  
1   2007-05-19  [Captain, Barbossa,, long, believed, to, be, d...  
2   2015-10-26  [A, cryptic, message, from, Bond’s, past, send...  
3   2012-07-16  [Following, the, death, of, District, Attorney...  
4   2012-03-07  [John, Carter, is, a, war-weary,, former, mili...

* Now we select only 'id', 'original_title', 'tags', release_date' columns and make a new dataset

In [39]:
df = movies[['id','original_title','tags','release_date']]
df

id                            original_title  \
0      19995                                    Avatar   
1        285  Pirates of the Caribbean: At World's End   
2     206647                                   Spectre   
3      49026                     The Dark Knight Rises   
4      49529                               John Carter   
...      ...                                       ...   
4804    9367                               El Mariachi   
4805   72766                                 Newlyweds   
4806  231617                 Signed, Sealed, Delivered   
4807  126186                          Shanghai Calling   
4808   25975                         My Date with Drew   

                                                   tags release_date  
0     [In, the, 22nd, century,, a, paraplegic, Marin...   2009-12-10  
1     [Captain, Barbossa,, long, believed, to, be, d...   2007-05-19  
2     [A, cryptic, message, from, Bond’s, past, send...   2015-10-26  
3     [Following, the, death, of, District, Attorney...   2012-07-16  
4     [John, Carter, is, a, war-weary,, former, mili...   2012-03-07  
...                                                 ...          ...  
4804  [El, Mariachi, just, wants, to, play, his, gui...   1992-09-04  
4805  [A, newlywed, couple's, honeymoon, is, upended...   2011-12-26  
4806  ["Signed,, Sealed,, Delivered", introduces, a,...   2013-10-13  
4807  [When, ambitious, New, York, attorney, Sam, is...   2012-05-03  
4808  [Ever, since, the, second, grade, when, he, fi...   2005-08-05  

[4805 rows x 4 columns]

* Now we convert list into string

In [40]:
df['tags'] = df['tags'].apply(lambda x:" ".join(x))
df

C:\Users\tejas\AppData\Local\Temp\ipykernel_8936\2524824924.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['tags'] = df['tags'].apply(lambda x:" ".join(x))


id                            original_title  \
0      19995                                    Avatar   
1        285  Pirates of the Caribbean: At World's End   
2     206647                                   Spectre   
3      49026                     The Dark Knight Rises   
4      49529                               John Carter   
...      ...                                       ...   
4804    9367                               El Mariachi   
4805   72766                                 Newlyweds   
4806  231617                 Signed, Sealed, Delivered   
4807  126186                          Shanghai Calling   
4808   25975                         My Date with Drew   

                                                   tags release_date  
0     In the 22nd century, a paraplegic Marine is di...   2009-12-10  
1     Captain Barbossa, long believed to be dead, ha...   2007-05-19  
2     A cryptic message from Bond’s past sends him o...   2015-10-26  
3     Following the death of District Attorney Harve...   2012-07-16  
4     John Carter is a war-weary, former military ca...   2012-03-07  
...                                                 ...          ...  
4804  El Mariachi just wants to play his guitar and ...   1992-09-04  
4805  A newlywed couple's honeymoon is upended by th...   2011-12-26  
4806  "Signed, Sealed, Delivered" introduces a dedic...   2013-10-13  
4807  When ambitious New York attorney Sam is sent t...   2012-05-03  
4808  Ever since the second grade when he first saw ...   2005-08-05  

[4805 rows x 4 columns]

now converting all strings into lowercase for better performance and understanding

In [41]:
df['tags'] = df['tags'].apply(lambda x:x.lower())

C:\Users\tejas\AppData\Local\Temp\ipykernel_8936\841579467.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['tags'] = df['tags'].apply(lambda x:x.lower())


In [42]:
df

id                            original_title  \
0      19995                                    Avatar   
1        285  Pirates of the Caribbean: At World's End   
2     206647                                   Spectre   
3      49026                     The Dark Knight Rises   
4      49529                               John Carter   
...      ...                                       ...   
4804    9367                               El Mariachi   
4805   72766                                 Newlyweds   
4806  231617                 Signed, Sealed, Delivered   
4807  126186                          Shanghai Calling   
4808   25975                         My Date with Drew   

                                                   tags release_date  
0     in the 22nd century, a paraplegic marine is di...   2009-12-10  
1     captain barbossa, long believed to be dead, ha...   2007-05-19  
2     a cryptic message from bond’s past sends him o...   2015-10-26  
3     following the death of district attorney harve...   2012-07-16  
4     john carter is a war-weary, former military ca...   2012-03-07  
...                                                 ...          ...  
4804  el mariachi just wants to play his guitar and ...   1992-09-04  
4805  a newlywed couple's honeymoon is upended by th...   2011-12-26  
4806  "signed, sealed, delivered" introduces a dedic...   2013-10-13  
4807  when ambitious new york attorney sam is sent t...   2012-05-03  
4808  ever since the second grade when he first saw ...   2005-08-05  

[4805 rows x 4 columns]

#### 8. Now we perform text vectorization

In [43]:
from sklearn.feature_extraction.text import CountVectorizer

In [44]:
cv = CountVectorizer(max_features = 5000, stop_words = 'english')

In [45]:
vectors = cv.fit_transform(df['tags']).toarray()
vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [46]:
cv.get_feature_names()  #these are the frequent words

C:\Users\tejas\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


['000',
 '007',
 '10',
 '100',
 '11',
 '12',
 '13',
 '14',
 '15',
 '150',
 '16',
 '17',
 '17th',
 '18',
 '18th',
 '19',
 '1930s',
 '1940s',
 '1944',
 '1950',
 '1950s',
 '1955',
 '1959',
 '1960',
 '1960s',
 '1962',
 '1964',
 '1965',
 '1967',
 '1970s',
 '1971',
 '1972',
 '1973',
 '1974',
 '1976',
 '1979',
 '1980',
 '1980s',
 '1985',
 '1997',
 '1999',
 '19th',
 '20',
 '200',
 '2000',
 '2003',
 '2009',
 '20th',
 '21st',
 '23',
 '24',
 '25',
 '30',
 '300',
 '3d',
 '40',
 '50',
 '500',
 '60',
 '60s',
 '70',
 '70s',
 'aaron',
 'aaroneckhart',
 'abandoned',
 'abandons',
 'abducted',
 'abilities',
 'ability',
 'able',
 'aboard',
 'abroad',
 'absence',
 'abuse',
 'abused',
 'abusive',
 'academic',
 'academy',
 'accept',
 'acceptance',
 'accepted',
 'accepts',
 'access',
 'accident',
 'accidental',
 'accidentally',
 'acclaimed',
 'accompanied',
 'accomplish',
 'accomplished',
 'account',
 'accountant',
 'accused',
 'ace',
 'achieve',
 'act',
 'acting',
 'action',
 'actions',
 'activist',
 'activi

In [47]:
len(cv.get_feature_names())

5000

#### 8. Applying steming technique for converting similar words in only one words.
* Ex : ['apply','appling','applied'] --> ['apply','apply','apply']

In [48]:
# !pip install nltk

In [49]:
import nltk
from nltk.stem.porter import PorterStemmer

In [50]:
ps = PorterStemmer()

In [51]:
def stem(text):
    a = list()
    for i in text.split():
        a.append(ps.stem(i))
    return " ".join(a)

In [ ]:
df['tags'] = df['tags'].apply(stem)

In [ ]:
df['tags']

#### 9. Now Calculating distance b/w movies in vector space (through angle)

In [ ]:
from sklearn.metrics.pairwise import  cosine_similarity

In [ ]:
total_similarity = cosine_similarity(vectors)
total_similarity.shape

#### 10. Making a function for recommend 5 movies

In [ ]:
def recommend(movie):
    movie_index = df[df['original_title'] == movie].index[0]
    distances = total_similarity[movie_index]
    movies_list = sorted(list(enumerate(distances)), reverse = True, key = lambda x:x[1])[1:6]
    
    for i in movies_list :
        print(df.iloc[i[0]].original_title)

In [ ]:
recommend('Avatar')

In [ ]:
#just for checking
df.iloc[151]

In [ ]:
# just for checking
df.iloc[151].original_title

#### 11. Dumping Pickle files

In [ ]:
import pickle

In [ ]:
pickle.dump(df,open('movies.pkl','wb'))

In [ ]:
pickle.dump(df.to_dict(),open('movies_dict.pkl','wb'))

In [ ]:
pickle.dump(total_similarity,open('similarity.pkl','wb'))

------------------------------------------------------------
------------------------------------------------------------
------------------------------------------------------------

---------------THANK YOU----------------